From 839898777226a3ed88c0859f25ffe712519b4ead Mon Sep 17 00:00:00 2001
From: Rocket Ma <marocketbd@gmail.com>
Date: Thu, 8 May 2026 00:00:00 +0000
Subject: [PATCH 0/0] stdio-common: Fix buffer overflow in scanf %mc

---
 stdio-common/vfscanf-internal.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/stdio-common/vfscanf-internal.c b/stdio-common/vfscanf-internal.c
index 59fc8208aa9..3d11ac261ed 100644
--- a/stdio-common/vfscanf-internal.c
+++ b/stdio-common/vfscanf-internal.c
@@ -855,8 +855,7 @@ __vfscanf_internal (FILE *s, const char *format, va_list argptr,
                        {
                          /* Enlarge the buffer.  */
                          size_t newsize
-                           = strsize
-                             + (strsize >= width ? width - 1 : strsize);
+                           = strsize + (strsize >= width ? width : strsize);
 
                          str = (char *) realloc (*strptr, newsize);
                          if (str == NULL)
@@ -929,7 +928,7 @@ __vfscanf_internal (FILE *s, const char *format, va_list argptr,
                      && wstr == (wchar_t *) *strptr + strsize)
                    {
                      size_t newsize
-                       = strsize + (strsize > width ? width - 1 : strsize);
+                       = strsize + (strsize >= width ? width : strsize);
                      /* Enlarge the buffer.  */
                      wstr = (wchar_t *) realloc (*strptr,
                                                  newsize * sizeof (wchar_t));
@@ -984,7 +983,7 @@ __vfscanf_internal (FILE *s, const char *format, va_list argptr,
                    && wstr == (wchar_t *) *strptr + strsize)
                  {
                    size_t newsize
-                     = strsize + (strsize > width ? width - 1 : strsize);
+                     = strsize + (strsize >= width ? width : strsize);
                    /* Enlarge the buffer.  */
                    wstr = (wchar_t *) realloc (*strptr,
                                                newsize * sizeof (wchar_t));
-- 
2.50.1
